LoopKit/DeviceManager/DeviceLog/PersistentDeviceLog.swift

```diff
@@ -0,0 +1,125 @@
+        persistentContainer.loadPersistentStores { description, error in
+            if let error = error {
+                fatalError("Unable to load persistent stores: \(error)")
+            }
+        }
+        managedObjectContext.persistentStoreCoordinator = persistentContainer.persistentStoreCoordinator
+    }
+    
+    public func log(deviceManagerIdentifier: String, deviceIdentifier: String?, type: DeviceLogEntryType, message: String, completion: ((Error?) -> Void)? = nil) {
+        managedObjectContext.perform {
+            let entry = DeviceLogEntry(context: self.managedObjectContext)
+            entry.deviceManager = deviceManagerIdentifier
+            entry.deviceIdentifier = deviceIdentifier
+            entry.type = type
+            entry.message = message
+            entry.timestamp = Date()
+            do {
+                try self.managedObjectContext.save()
+                self.log.default("Logged: %{public}@ (%{public}@) %{public}@", String(describing: type), deviceIdentifier ?? "", message)
+            } catch let error {
+                self.log.error("Could not store device log entry %{public}@", String(describing: error))
+                completion?(error)
+            }
+            completion?(nil)
+        }
+    }
+    
+    public func getLogEntries(startDate: Date, endDate: Date? = nil, completion: @escaping (_ result: Result<[StoredDeviceLogEntry], Error>) -> Void) {
+        
+        managedObjectContext.perform {
+            var predicate: NSPredicate = NSPredicate(format: "timestamp >= %@", startDate as NSDate)
+            if let endDate = endDate {
+                let endFilter = NSPredicate(format: "timestamp < %@", endDate as NSDate)
+                predicate = NSCompoundPredicate(andPredicateWithSubpredicates: [predicate, endFilter])
+            }
+            
+            let request: NSFetchRequest<DeviceLogEntry> = DeviceLogEntry.fetchRequest()
+            request.predicate = predicate
+            request.sortDescriptors = [NSSortDescriptor(key: "timestamp", ascending: true)]
+            
+            do {
+                let entries = try self.managedObjectContext.fetch(request)
+                completion(.success(entries.map { StoredDeviceLogEntry(managedObject: $0) } ))
+                self.purgeExpiredLogEntries()
+            } catch let error {
+                completion(.failure(error))
+            }
+        }
+    }
+    
+    // Should only be called from managed object context queue
+    private func purgeExpiredLogEntries() {
+        let predicate = NSPredicate(format: "timestamp < %@", earliestLogDate as NSDate)
+
+        do {
+            let fetchRequest: NSFetchRequest<DeviceLogEntry> = DeviceLogEntry.fetchRequest()
+            fetchRequest.predicate = predicate
+            let count = try managedObjectContext.deleteObjects(matching: fetchRequest)
+            self.log.info("Deleted %d DeviceLogEntries", count)
```

Comment:
[nit] I don't believe `self.` prefix is needed here. Just to be consistent with `log.error` below.